src/zircon/metadatacache/leasing/leasing.go

```diff
@@ -0,0 +1,271 @@
+package leasing
+
+import (
+	"zircon/metadatacache/access"
+	"zircon/apis"
+	"zircon/rpc"
+	"sync"
+	"github.com/pkg/errors"
+	"fmt"
+	"time"
+)
+
+type Lease struct {
+	// TODO: lease-level locking
+	Version  apis.Version
+	Contents []byte
+}
+
+type Leasing struct {
+	access *access.Access
+	etcd   apis.EtcdInterface
+
+	mu        sync.Mutex
+	cancel    chan struct{}
+	done      chan struct{}
+	safe      bool
+	leases    map[apis.MetadataID]*Lease
+}
+
+func ConstructLeasing(etcd apis.EtcdInterface, cache rpc.ConnectionCache) (*Leasing, error) {
+	chunkAccess, err := access.ConstructAccess(etcd, cache)
+	if err != nil {
+		return nil, err
+	}
+	return &Leasing{
+		access: chunkAccess,
+		etcd: etcd,
+		leases: make(map[apis.MetadataID]*Lease),
+	}, nil
+}
+
+func (l *Leasing) Start() error {
+	l.mu.Lock()
+	defer l.mu.Unlock()
+
+	if l.cancel != nil {
+		return errors.New("cannot start leasing agent; already running!")
+	}
+	if l.done != nil {
+		return errors.New("cannot start leasing agent; not done stopping!")
+	}
+
+	err := l.etcd.BeginMetadataLease()
+	if err != nil {
+		return err
+	}
+
+	l.cancel = make(chan struct{})
+	l.done = make(chan struct{})
+	l.safe = true
+
+	go l.mainloop()
+
+	return l.ensureRenewed_LK()
+}
+
+func (l *Leasing) Stop() error {
+	done := func() chan struct{} {
+		l.mu.Lock()
+		defer l.mu.Unlock()
+		if l.cancel == nil {
+			return nil
+		}
+		close(l.cancel)
+		l.cancel = nil
+		if l.done == nil {
+			panic("done should not be nil yet!")
+		}
+		return l.done
+	}()
+	if done == nil {
+		return errors.New("either already stopped or already in the process of stopping!")
+	}
+	<-done
+	l.mu.Lock()
+	defer l.mu.Unlock()
+	l.done = nil
+	return nil
+}
+
+func (l *Leasing) mainloop() {
+	defer func() {
+		close(l.done)
+	}()
+	for {
+		select {
+		case <-l.cancel:
+			break
+		case <-time.After(l.etcd.GetMetadataLeaseTimeout() / 3):
+			err := l.etcd.RenewMetadataClaims()
+			if err != nil {
+				l.notifyUnsafe()
+				break
+			}
+		}
+	}
+}
+
+func (l *Leasing) notifyUnsafe() {
+	l.mu.Lock()
+	defer l.mu.Unlock()
+	l.notifyUnsafe_LK()
+}
+
+func (l *Leasing) ensureRenewed_LK() error {
+	if err := l.etcd.RenewMetadataClaims(); err != nil {
+		l.notifyUnsafe_LK()
+		return err
+	}
+	return nil
+}
+
+func (l *Leasing) notifyUnsafe_LK() {
+	l.safe = false
+	l.leases = make(map[apis.MetadataID]*Lease)
+}
+
+// Get *any* unleased block. If everything that exists is leased, create a new block with all zeroes and lease it.
```

Comment:
Maybe use a name that hints a bit more that it will allocate new leases if it can't find one